File: examples/python/e50hx.py
#!/usr/bin/python
# Author: Jon Trulson <jtrulson@ics.com>
# Copyright (c) 2016 Intel Corporation.
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

from __future__ import print_function
import time, sys, signal, atexit
from upm import pyupm_e50hx as sensorObj

def main():
    ## Exit handlers ##
    # This function stops python from printing a stacktrace when you hit control-C
    def SIGINTHandler(signum, frame):
        raise SystemExit

    # This function lets you run code on exit
    def exitHandler():
        print("Exiting...")
        sys.exit(0)

    # Register exit handlers
    atexit.register(exitHandler)
    signal.signal(signal.SIGINT, SIGINTHandler)

    # You will need to edit this example to conform to your site and your
    # devices, specifically the Device Object Instance ID passed to the
    # constructor, and the arguments to initMaster() that are
    # appropriate for your BACnet network.

    defaultDev = "/dev/ttyUSB0"

    # if an argument was specified, use it as the device instead
    if (len(sys.argv) > 1):
        defaultDev = sys.argv[1]

    print("Using device", defaultDev)
    print("Initializing...")

    # Instantiate an E50HX object for an E50HX device that has 1075425
    # as it's unique Device Object Instance ID.  NOTE: You will
    # certainly want to change this to the correct value for your
    # device(s).
    sensor = sensorObj.E50HX(1075425)

    # Initialize our BACnet master, if it has not already been
    # initialized, with the device and baudrate, choosing 1000001 as
    # our unique Device Object Instance ID, 2 as our MAC address and
    # using default values for maxMaster and maxInfoFrames
    sensor.initMaster(defaultDev, 38400, 1000001, 2)

    # Uncomment to enable debugging output
    # sensor.setDebug(True);

    # output the serial number and firmware revision
    print()
    print("Device Description:", sensor.getDeviceDescription())
    print("Device Location:", sensor.getDeviceLocation())
    print()

    # update and print available values every second
    while (1):
        print("System Voltage:", end=' ')
        print(sensor.getAnalogValue(sensorObj.E50HX.AV_System_Voltage), end=' ')
        print(" ", end=' ')
        print(sensor.getAnalogValueUnits(sensorObj.E50HX.AV_System_Voltage))

        print("System Type:", end=' ')
        print(sensor.getAnalogValue(sensorObj.E50HX.AV_System_Type))

        print("Energy Consumption:", end=' ')
        print(sensor.getAnalogInput(sensorObj.E50HX.AI_Energy), end=' ')
        print(" ", end=' ')
        print(sensor.getAnalogInputUnits(sensorObj.E50HX.AI_Energy))

        print("Power Up Counter:", end=' ')
        print(sensor.getAnalogInput(sensorObj.E50HX.AI_Power_Up_Count))

        print()
        time.sleep(5)

if __name__ == '__main__':
    main()

File: upm/src/e50hx/e50hx.cxx
/*
 * Author: Jon Trulson <jtrulson@ics.com>
 * Copyright (c) 2016 Intel Corporation.
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#include <unistd.h>
#include <assert.h>
#include <errno.h>
#include <iostream>
#include <stdexcept>
#include <string>

#include "e50hx.hpp"

using namespace upm;
using namespace std;


E50HX::E50HX(uint32_t targetDeviceObjectID) :
  BACNETUTIL(targetDeviceObjectID)
{
  setDebug(false);

  // we disable this by default for performance reasons
  checkReliability(false);
}

E50HX::~E50HX()
{
}

uint16_t E50HX::getAlarmBits()
{
  return uint16_t(getAnalogInput(AI_Alarm_Bitmap));
}

void E50HX::writeConfig(CFG_VALUES_T config)
{
  setAnalogValue(AV_Config, float(config));
}

void E50HX::writeSystemType(SYSTEM_TYPES_T systype)
{
  setAnalogValue(AV_System_Type, float(systype));
}

void E50HX::writeCTRatioPrimary(float ctRatio)
{
  if (ctRatio < 5 || ctRatio > 32000)
    {
      throw std::out_of_range(std::string(__FUNCTION__)
                              + ": ctRatio must be between 5-32000");
    }

  setAnalogValue(AV_CT_Ratio_Primary, ctRatio);
}

void E50HX::writeCTRatioSecondary(CT_SECONDARY_T ctRatio)
{
  setAnalogValue(AV_CT_Ratio_Secondary, ctRatio);
}

void E50HX::writePTRatio(float ptRatio)
{
  if (ptRatio < 0.01 || ptRatio > 320.0)
    {
      throw std::out_of_range(std::string(__FUNCTION__)
                              + ": ptRatio must be between 0.01-320.0");
    }

  setAnalogValue(AV_PT_Ratio, ptRatio);
}

void E50HX::writeSystemVoltage(float sysVolts)
{
  if (sysVolts < 82.0 || sysVolts > 32000.0)
    {
      throw std::out_of_range(std::string(__FUNCTION__)
                              + ": sysVolts must be between 82.0-32000.0");
    }

  setAnalogValue(AV_System_Voltage, sysVolts);
}

void E50HX::writeDisplayUnits(DISP_UNITS_T dispUnits)
{
  setAnalogValue(AV_Display_Units, dispUnits);
}

void E50HX::writePhaseLossVT(float phaseLoss)
{
  if (phaseLoss < 1.0 || phaseLoss > 99.0)
    {
      throw std::out_of_range(std::string(__FUNCTION__)
                              + ": phaseLoss must be between 1.0-99.0");
    }

  setAnalogValue(AV_Phase_Loss_Voltage_Threshold, phaseLoss);
}

void E50HX::writePhaseLossIT(float phaseLoss)
{
  if (phaseLoss < 1.0 || phaseLoss > 99.0)
    {
      throw std::out_of_range(std::string(__FUNCTION__)
                              + ": phaseLoss must be between 1.0-99.0");
    }

  setAnalogValue(AV_Phase_Loss_Imbalance_Threshold, phaseLoss);
}
